### DISCUSSION TITLE
Passing Type Information in OpenDDS RTPS Packets

### DISCUSSION THREAD

**Discussion by Idan-Neeman** [OpenDDS/OpenDDS#4212] [Q&A]
I am wondering if there is a way to configure OpenDDS to include typecodes (PID_TYPECODE) or typeobjects within the RTPS packets during the Endpoint Discovery phase.
Preferably by configuration that allows me to change that before the application is running, instead of compiling the code.

**Comment by mitza-oci** [comment]
Type objects are exchanged as part of the type lookup service messages (if needed), which are part of endpoint discovery.  This is defined by the DDS-XTypes spec (v1.3).

**Reply by Idan-Neeman** [reply]
I captured the data on Wireshark, but I don't have the type object there. I'm attaching a screenshot of data(w):
![DATA W](https://github.com/OpenDDS/OpenDDS/assets/66280125/1ce14508-52a9-4b67-ab26-7ec55da505e6)

**Reply by mitza-oci** [reply]
This is by design.  The protocol is not going to send what's not needed.

Try starting a subscriber that uses dynamic types or doesn't already have the typeobject (OpenDDS includes a dynamic subscriber in `tools/inspect`).

**Reply by Idan-Neeman** [reply]
Is there a way to force the sending of the type object anyway? I mean, even if the subscriber already has the type object in advance.
Thank you!

**Reply by mitza-oci** [reply]
Running the inspect tool is one way to force sending the type object.  Also see comment below by @jrw972

**Reply by Idan-Neeman** [reply]
Let's say according to the example, should I run it this way: "inspect.exe Messenger.Message Message 42"?
Because if so, there is still no type object.

UPDATE:
It requires the module name too, therefore:
"Movie Discussion List" "Messenger::Message" 42

**Reply by mitza-oci** [reply]
Anything on the built-in endpoint for type lookup service replies?  The writer Entity Id is `{{00, 03, 01}, c3}` (Table 61 in XTypes)

**Reply by Idan-Neeman** [reply]
No, nothing.
By the way, when running inspect, I got this warning: "(18256|22068) WARNING: DynamicDataBase: Passed null DynamicType pointer".

I guess because the type object was not passed, the reader could not build the DynamicData.

**Reply by iguessthislldo** [reply]
> By the way, when running inspect, I got this error: "(18256|22068) WARNING: DynamicDataBase: Passed null DynamicType pointer".

The code should be checking for that case, otherwise it could crash trying to use the invalid DynamicData, so that's something that should probably be fixed. However I played around with inspect and I only get that message if I don't change the IDL to build with `-Gxtypes-complete`. Are you doing both that and enabling complete type objects like mentioned in the inspect README?

**Reply by Idan-Neeman** [reply]
Thank you!
The inspect tool is working well so far. However, I'm still not able to see the PID_TYPECODE or PID_TYPE_OBJECT fields in the data[w] packets when viewing them in Wireshark. I thought these fields should be present based on the documentation, but they do not appear in the captures I've checked. Please let me know if you have any insights into why these particular protocol fields are not being populated as expected in the packets.

**Comment by jrw972** [comment]
One could add a run-time option for this.  Let us know if you are interested in contributing a feature like this.

**Comment by mitza-oci** [comment]
Wireshark doesn't decode it (unless that's changed recently). Check for a
parameter with PID 0x75 for type info and then keep in mind that the
message what Wireshark calls DATA[w] will never have type objects, just
type identifiers. Type objects are in the reply message on the type lookup
service built-in endpoint.

**Reply by Idan-Neeman** [reply]
When using the RTI Connext implementation, I noticed that the data structure (type object) is passed as part of the data[w]. From your reply my understanding is that embedding type information in data[w] is not formally part of the DDSI-RTPS standard. Is this an extension specific to RTI's implementation?

Additionally, I'm wondering which submessage ID field contains the type object response after a lookup. I've checked the whole packets before sending data but don't see the type object data I expected based on the lookup. If you have any pointers on where to find this data, I'd appreciate your guidance.

We are working on an interoperable DDS packet dissector and seeking advice on the best way to achieve broad vendor compatibility. Are there any essential proprietary extensions or assumptions I should avoid? Your guidance on the most effective approach would be greatly appreciated.

**Reply by jrw972** [reply]
If you want to see it in action, the XTypes test in OpenDDS engages the TypeLookupService.  The following filter worked for me `((rtps) && (rtps.sm.id == 0x15)) && (rtps.sm.wrEntityId == 0x000301c3)`.

From the discussion, one would need to consider at least two levels of support.  Participants using XTypes 1.2 and 1.3 should have TypeLookupService endpoints, and one would need to use them to acquire the type information.  These can be detected in the SPDP message in the bitset that indicates supported endpoints.  If these bits are set, one could conclude that the remote participant supports 1.2 or 1.3.  (I'm unaware of any message explicitly describing the version of XTypes supported.)

For pre-1.2 participants, the TypeObject should be in the publication and subscription data as you have already discovered.

**Comment by mitza-oci** [comment]
My guess is that the RTI example is using an old version of the XTypes spec.

I'm not in front of my computer now, just using my phone. You can find all
the details of how this works in the spec.

**Reply by Idan-Neeman** [reply]
This means that if RTI implemented XTypes in an older version (1.2 for example), does this mean that in the implementation of OpenDDS that works with XTypes 1.3 there is no backward compatibility support?
I see that the inspect tool fails to dynamically listen to the RTI shape demo tool.

**Reply by jrw972** [reply]
The TypeLookupService was introduced in XTypes 1.2 so the RTI shape demo tool that you are working with is probably from XTypes 1.1 or earlier.  I don't see a reasonable way for a pre-TypeLookupService participant to interoperate with a participant that is using the TypeLookupService as the 1.2/1.3 participant would have to specially encode publication and subscription data to contain the additional parameters.

**Reply by Idan-Neeman** [reply]
If one DDS implementation uses XTypes 1.1 (for example old version of RTI connext) and another uses XTypes 1.2/1.3 (opendds), will the mismatch prevent a dynamic subscriber from interpreting and reading the data properly?

**Reply by jrw972** [reply]
Yes, assuming a type mismatch.  The type information from the 1.1 participant will be silently ignored as the 1.3 participant does not understand those PIDs.  Since the 1.1 participant does not have TypeLookupService endpoints, the 1.3 participant will not attempt to do a type lookup or match.  The type information will be absent in information sent from the 1.3 participant to the 1.1 participant.  Thus, I speculate that the 1.1 participant would fail to match as well.

Given a high enough log in OpenDDS, one should see the discovery of the remote publisher/subscriber but no ensuing match.

**Reply by mitza-oci** [reply]
From a spec/interop point of view, please see section 2.3 "Optional XTYPES 1.1 Interoperability Profile" and https://opendds.readthedocs.io/en/latest-release/devguide/xtypes.html#unimplemented-features

**Reply by Idan-Neeman** [reply]
Is this updated information? I noticed that the annotation @optional is listed as not being supported. Is this really the case?

**Reply by mitza-oci** [reply]
The information is correct as far as I know. `@optional` has been covered (probably?) multiple times in various GitHub issues/discussions.
